File: pushfill/display.py
# ────────────────────────────────────────────────────────────────────────────────────────
#   display.py
#   ──────────
#
#   Live terminal status display for pushfill. Uses Unicode box-drawing
#   characters and ANSI escape sequences for a polished, framed display
#   showing speed, progress, disk usage, and ETA.
#
#   (c) 2026 WaterJuice — Unlicense; see LICENSE in the project root.
#
#   Authors
#   ───────
#   bena (via Claude)
# ────────────────────────────────────────────────────────────────────────────────────────

# ────────────────────────────────────────────────────────────────────────────────────────
#   Imports
# ────────────────────────────────────────────────────────────────────────────────────────

import re
import shutil
import sys
import time
from typing import Callable
from typing import Optional
from pushfill.colour import bold
from pushfill.colour import cyan
from pushfill.colour import dim
from pushfill.colour import green
from pushfill.colour import magenta
from pushfill.colour import yellow

# ────────────────────────────────────────────────────────────────────────────────────────
#   Helpers
# ────────────────────────────────────────────────────────────────────────────────────────

EMA_ALPHA = 0.1  # smoothing factor — lower = slower, smoother rolling average


# ────────────────────────────────────────────────────────────────────────────────────────
def format_size(nbytes: int) -> str:
    """Format a byte count as a human-readable string."""
    if nbytes < 1024:
        return f"{nbytes} B"
    elif nbytes < 1024 * 1024:
        return f"{nbytes / 1024:.1f} KB"
    elif nbytes < 1024 * 1024 * 1024:
        return f"{nbytes / (1024 * 1024):.1f} MB"
    elif nbytes < 1024 * 1024 * 1024 * 1024:
        return f"{nbytes / (1024 * 1024 * 1024):.2f} GB"
    else:
        return f"{nbytes / (1024 * 1024 * 1024 * 1024):.2f} TB"


# ────────────────────────────────────────────────────────────────────────────────────────
def format_time(seconds: float) -> str:
    """Format seconds as h:mm:ss or m:ss."""
    seconds = max(0, int(seconds))
    if seconds < 3600:
        return f"{seconds // 60}:{seconds % 60:02d}"
    return f"{seconds // 3600}:{(seconds % 3600) // 60:02d}:{seconds % 60:02d}"


# ────────────────────────────────────────────────────────────────────────────────────────
#   Display Class
# ────────────────────────────────────────────────────────────────────────────────────────


# ────────────────────────────────────────────────────────────────────────────────────────
class Display:
    """Live terminal display with box-drawn framing."""

    def __init__(
        self,
        target_path: str,
        target_size: Optional[int] = None,
        goal_bytes: Optional[int] = None,
        num_workers: int = 1,
        version: str = "dev",
    ) -> None:
        self._target_path = target_path
        self._target_size = target_size
        self._goal_bytes = goal_bytes
        self._num_workers = num_workers
        self._version = version
        self._start_time = time.monotonic()
        self._prev_total = 0
        self._prev_time = self._start_time
        self._ema_rate: Optional[float] = None
        self._lines_printed = 0

    # ──────────────────────────────────────────────────────────────────────────────────
    def set_goal(self, goal_bytes: Optional[int]) -> None:
        """Update the goal bytes (e.g. when purgeable space is reclaimed)."""
        self._goal_bytes = goal_bytes

    # ──────────────────────────────────────────────────────────────────────────────────
    def _move_up_and_clear(self) -> None:
        """Move cursor up to overwrite previous output."""
        if self._lines_printed > 0:
            sys.stdout.write(f"\033[{self._lines_printed}A\033[J")

    # ──────────────────────────────────────────────────────────────────────────────────
    def _get_bar(
        self,
        fraction: float,
        width: int,
        colour_fn: Optional[Callable[[str], str]] = None,
    ) -> str:
        """Render a progress bar with the given colour."""
        fraction = max(0.0, min(1.0, fraction))
        filled = int(fraction * width)
        empty = width - filled
        if colour_fn is None:
            colour_fn = green
        bar_fill = colour_fn("\u2588" * filled)
        bar_empty = dim("\u2591" * empty)
        return f"{bar_fill}{bar_empty}"

    # ──────────────────────────────────────────────────────────────────────────────────
    def _box_line(self, content: str, inner_w: int) -> str:
        """Wrap content in box borders, padding to inner_w visible characters."""
        visible = _strip_ansi(content)
        pad = inner_w - len(visible)
        if pad < 0:
            pad = 0
        border = dim("\u2502")
        return f"{border}  {content}{' ' * pad}  {border}"

    # ──────────────────────────────────────────────────────────────────────────────────
    def _box_top(self, inner_w: int) -> str:
        return dim("\u256d" + "\u2500" * (inner_w + 4) + "\u256e")

    def _box_sep(self, inner_w: int) -> str:
        return dim("\u251c" + "\u2500" * (inner_w + 4) + "\u2524")

    def _box_bottom(self, inner_w: int) -> str:
        return dim("\u2570" + "\u2500" * (inner_w + 4) + "\u256f")

    # ──────────────────────────────────────────────────────────────────────────────────
    def update(self, total_bytes: int) -> None:
        """Update the display with current progress (8 lines max)."""
        now = time.monotonic()
        elapsed = now - self._start_time
        dt = now - self._prev_time

        # Calculate rates
        if dt > 0:
            instant_rate = (total_bytes - self._prev_total) / dt
        else:
            instant_rate = 0.0

        # EMA smoothing (slow rolling average)
        if self._ema_rate is None:
            self._ema_rate = instant_rate
        else:
            self._ema_rate = EMA_ALPHA * instant_rate + (1 - EMA_ALPHA) * self._ema_rate

        if elapsed > 0:
            avg_rate = total_bytes / elapsed
        else:
            avg_rate = 0.0

        self._prev_total = total_bytes
        self._prev_time = now

        # Terminal width → inner box width
        term_width = shutil.get_terminal_size((80, 24)).columns
        inner_w = max(46, min(term_width - 6, 72))
        bar_width = max(10, inner_w - 10)

        # Build lines
        self._move_up_and_clear()
        lines: list[str] = []

        # Line 1: top border
        lines.append(self._box_top(inner_w))

        # Line 2: title + elapsed
        elapsed_str = format_time(elapsed)
        title_left = bold(cyan(f"pushfill {self._version}"))
        title_right = dim(f"Elapsed {elapsed_str}")
        title_pad = inner_w - _visible_len(title_left) - _visible_len(title_right)
        border = dim("\u2502")
        lines.append(
            f"{border}  {title_left}{' ' * max(1, title_pad)}{title_right}  {border}"
        )

        # Line 3: separator
        lines.append(self._box_sep(inner_w))

        # Line 4: speed (rolling average)
        ema = self._ema_rate or 0.0
        speed_mbs = ema / 1e6
        speed_gbps = ema * 8 / 1e9
        lines.append(
            self._box_line(
                f"{magenta(bold('Speed'))}     {cyan(f'{speed_mbs:,.1f} MB/s')}   "
                f"{dim(f'({speed_gbps:.2f} Gbps)')}",
                inner_w,
            )
        )

        # Line 5: average (total / elapsed)
        avg_mbs = avg_rate / 1e6
        avg_gbps = avg_rate * 8 / 1e9
        lines.append(
            self._box_line(
                f"{magenta(bold('Average'))}   {cyan(f'{avg_mbs:,.1f} MB/s')}   "
                f"{dim(f'({avg_gbps:.2f} Gbps)')}",
                inner_w,
            )
        )

        # Line 6: written + ETA (or disk %)
        if self._goal_bytes is not None and self._goal_bytes > 0:
            progress = min(total_bytes / self._goal_bytes, 1.0)
            size_str = f"{format_size(total_bytes)} / {format_size(self._goal_bytes)}"
            if avg_rate > 0 and progress < 1.0:
                eta_str = (
                    f"ETA {format_time((self._goal_bytes - total_bytes) / avg_rate)}"
                )
            else:
                eta_str = ""
            left = f"{magenta(bold('Written'))}   {cyan(size_str)}"
            right = yellow(eta_str) if eta_str else ""
        else:
            progress = 0.0
            left = f"{magenta(bold('Written'))}   {cyan(format_size(total_bytes))}"
            try:
                usage = shutil.disk_usage(self._target_path)
                disk_pct = usage.used / usage.total * 100 if usage.total > 0 else 0.0
                progress = disk_pct / 100.0
                right = dim(f"Disk {disk_pct:.1f}%")
            except OSError:
                right = ""

        pad = inner_w - _visible_len(left) - _visible_len(right)
        border = dim("\u2502")
        lines.append(f"{border}  {left}{' ' * max(1, pad)}{right}  {border}")

        # Line 7: progress bar
        pct_str = f"{progress * 100:.1f}%"
        prog_bar = self._get_bar(progress, bar_width, green)
        lines.append(self._box_line(f"{prog_bar}  {green(pct_str)}", inner_w))

        # Line 8: bottom border
        lines.append(self._box_bottom(inner_w))

        output = "\n".join(lines) + "\n"
        sys.stdout.write(output)
        sys.stdout.flush()
        self._lines_printed = len(lines)

    # ──────────────────────────────────────────────────────────────────────────────────
    def final_report(self, total_bytes: int, interrupted: bool = False) -> None:
        """Print the final summary."""
        elapsed = time.monotonic() - self._start_time
        avg_rate = total_bytes / elapsed if elapsed > 0 else 0.0
        avg_mbs = avg_rate / 1e6
        avg_gbps = avg_rate * 8 / 1e9

        if interrupted:
            # Don't clear the live display — append summary below it
            sys.stdout.write(
                f"\n  {yellow('Interrupted')} — wrote {cyan(format_size(total_bytes))} "
                f"in {bold(format_time(elapsed))} "
                f"({avg_mbs:,.1f} MB/s, {avg_gbps:.2f} Gbps) "
                f"across {self._num_workers} workers\n"
            )
            sys.stdout.flush()
            return

        # Normal completion — replace live display with summary box
        self._move_up_and_clear()
        self._lines_printed = 0

        term_width = shutil.get_terminal_size((80, 24)).columns
        inner_w = max(46, min(term_width - 6, 72))

        lines: list[str] = []
        lines.append(self._box_top(inner_w))
        lines.append(
            self._box_line(
                f"{bold(cyan(f'pushfill {self._version}'))}  {green('Done')} — "
                f"wrote {cyan(format_size(total_bytes))} "
                f"in {bold(format_time(elapsed))}",
                inner_w,
            )
        )
        lines.append(
            self._box_line(
                f"{magenta('Average:')} {cyan(f'{avg_mbs:,.1f} MB/s')} "
                f"{dim(f'({avg_gbps:.2f} Gbps)')} "
                f"across {self._num_workers} workers",
                inner_w,
            )
        )
        lines.append(self._box_bottom(inner_w))

        output = "\n".join(lines) + "\n"
        sys.stdout.write(output)
        sys.stdout.flush()


# ────────────────────────────────────────────────────────────────────────────────────────
#   ANSI Helpers
# ────────────────────────────────────────────────────────────────────────────────────────

_ANSI_RE = re.compile(r"\033\[[0-9;]*m")


def _strip_ansi(s: str) -> str:
    """Remove ANSI escape codes from a string."""
    return _ANSI_RE.sub("", s)


def _visible_len(s: str) -> int:
    """Return the visible length of a string (excluding ANSI codes)."""
    return len(_strip_ansi(s))

File: internal/display.go
// ---------------------------------------------------------------------------------------
//
//	display.go
//	----------
//
//	Live terminal status display for pushfill. Uses Unicode box-drawing
//	characters and ANSI escape sequences for a polished, framed display
//	showing speed, progress, disk usage, and ETA.
//
//	Replicates the exact output of the Python version's Display class.
//
//	(c) 2026 WaterJuice — Released under the Unlicense; see LICENSE.
//
//	Authors
//	-------
//	bena (via Claude)
//
//	Version History
//	---------------
//	Feb 2026 - Created (Python)
//	Mar 2026 - Rewritten in Go
//
// ---------------------------------------------------------------------------------------
package internal

// ---------------------------------------------------------------------------------------
//
//	Imports
//
// ---------------------------------------------------------------------------------------

import (
	"fmt"
	"os"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------------------
//
//	Constants
//
// ---------------------------------------------------------------------------------------

const emaAlpha = 0.1 // smoothing factor — lower = slower, smoother rolling average

// ---------------------------------------------------------------------------------------
//
//	Display
//
// ---------------------------------------------------------------------------------------

// ---------------------------------------------------------------------------------------
// Display renders live terminal output with box-drawn framing.
type Display struct {
	targetPath   string
	targetSize   int64 // 0 = fill disk
	goalBytes    int64 // 0 = unknown
	numWorkers   int
	version      string
	startTime    time.Time
	prevTotal    int64
	prevTime     time.Time
	emaRate      float64
	emaInit      bool
	linesPrinted int
}

// ---------------------------------------------------------------------------------------
// NewDisplay creates a new Display instance.
func NewDisplay(targetPath string, targetSize int64, goalBytes int64, numWorkers int, version string) *Display {
	now := time.Now()
	return &Display{
		targetPath: targetPath,
		targetSize: targetSize,
		goalBytes:  goalBytes,
		numWorkers: numWorkers,
		version:    version,
		startTime:  now,
		prevTime:   now,
	}
}

// ---------------------------------------------------------------------------------------
// SetGoal updates the goal bytes (e.g. when purgeable space is reclaimed).
func (d *Display) SetGoal(goalBytes int64) {
	d.goalBytes = goalBytes
}

// ---------------------------------------------------------------------------------------
// Update refreshes the display with current progress.
func (d *Display) Update(totalBytes int64) {
	tty := isStdoutTerminal()
	now := time.Now()
	elapsed := now.Sub(d.startTime).Seconds()
	dt := now.Sub(d.prevTime).Seconds()

	// Calculate rates
	var instantRate float64
	if dt > 0 {
		instantRate = float64(totalBytes-d.prevTotal) / dt
	}

	// EMA smoothing
	if !d.emaInit {
		d.emaRate = instantRate
		d.emaInit = true
	} else {
		d.emaRate = emaAlpha*instantRate + (1-emaAlpha)*d.emaRate
	}

	var avgRate float64
	if elapsed > 0 {
		avgRate = float64(totalBytes) / elapsed
	}

	d.prevTotal = totalBytes
	d.prevTime = now

	// Terminal width → inner box width
	termWidth := getTerminalWidth()
	innerW := termWidth - 6
	if innerW < 46 {
		innerW = 46
	}
	if innerW > 72 {
		innerW = 72
	}
	barWidth := innerW - 10
	if barWidth < 10 {
		barWidth = 10
	}

	// Move up and clear previous output
	d.moveUpAndClear()
	var lines []string

	// Line 1: top border
	lines = append(lines, d.boxTop(innerW, tty))

	// Line 2: title + elapsed
	elapsedStr := formatTime(elapsed)
	titleLeft := colorBold(colorCyan("pushfill "+d.version, tty), tty)
	titleRight := colorDim("Elapsed "+elapsedStr, tty)
	titlePad := innerW - visibleLen(titleLeft) - visibleLen(titleRight)
	if titlePad < 1 {
		titlePad = 1
	}
	border := colorDim("│", tty)
	lines = append(lines, fmt.Sprintf("%s  %s%s%s  %s", border, titleLeft, strings.Repeat(" ", titlePad), titleRight, border))

	// Line 3: separator
	lines = append(lines, d.boxSep(innerW, tty))

	// Line 4: speed (rolling average)
	ema := d.emaRate
	speedMBs := ema / 1e6
	speedGbps := ema * 8 / 1e9
	lines = append(lines, d.boxLine(
		fmt.Sprintf("%s     %s   %s",
			colorMagenta(colorBold("Speed", tty), tty),
			colorCyan(formatFloat1(speedMBs)+" MB/s", tty),
			colorDim(fmt.Sprintf("(%.2f Gbps)", speedGbps), tty)),
		innerW, tty))

	// Line 5: average
	avgMBs := avgRate / 1e6
	avgGbps := avgRate * 8 / 1e9
	lines = append(lines, d.boxLine(
		fmt.Sprintf("%s   %s   %s",
			colorMagenta(colorBold("Average", tty), tty),
			colorCyan(formatFloat1(avgMBs)+" MB/s", tty),
			colorDim(fmt.Sprintf("(%.2f Gbps)", avgGbps), tty)),
		innerW, tty))

	// Line 6: written + ETA
	var progress float64
	var left, right string
	if d.goalBytes > 0 {
		progress = float64(totalBytes) / float64(d.goalBytes)
		if progress > 1.0 {
			progress = 1.0
		}
		sizeStr := fmt.Sprintf("%s / %s", formatSize(totalBytes), formatSize(d.goalBytes))
		left = fmt.Sprintf("%s   %s", colorMagenta(colorBold("Written", tty), tty), colorCyan(sizeStr, tty))
		if avgRate > 0 && progress < 1.0 {
			eta := float64(d.goalBytes-totalBytes) / avgRate
			right = colorYellow(fmt.Sprintf("ETA %s", formatTime(eta)), tty)
		}
	} else {
		left = fmt.Sprintf("%s   %s", colorMagenta(colorBold("Written", tty), tty), colorCyan(formatSize(totalBytes), tty))
		du := getDiskUsage(d.targetPath)
		if du.total > 0 {
			diskPct := float64(du.used) / float64(du.total) * 100
			progress = diskPct / 100.0
			right = colorDim(fmt.Sprintf("Disk %.1f%%", diskPct), tty)
		}
	}
	pad := innerW - visibleLen(left) - visibleLen(right)
	if pad < 1 {
		pad = 1
	}
	lines = append(lines, fmt.Sprintf("%s  %s%s%s  %s", border, left, strings.Repeat(" ", pad), right, border))

	// Line 7: progress bar
	pctStr := fmt.Sprintf("%.1f%%", progress*100)
	progBar := d.getBar(progress, barWidth, tty)
	lines = append(lines, d.boxLine(fmt.Sprintf("%s  %s", progBar, colorGreen(pctStr, tty)), innerW, tty))

	// Line 8: bottom border
	lines = append(lines, d.boxBottom(innerW, tty))

	output := strings.Join(lines, "\n") + "\n"
	fmt.Fprint(os.Stdout, output)
	d.linesPrinted = len(lines)
}

// ---------------------------------------------------------------------------------------
// FinalReport prints the final summary.
func (d *Display) FinalReport(totalBytes int64, interrupted bool) {
	tty := isStdoutTerminal()
	elapsed := time.Since(d.startTime).Seconds()
	var avgRate float64
	if elapsed > 0 {
		avgRate = float64(totalBytes) / elapsed
	}
	avgMBs := avgRate / 1e6
	avgGbps := avgRate * 8 / 1e9

	if interrupted {
		fmt.Fprintf(os.Stdout, "\n  %s — wrote %s in %s (%s MB/s, %.2f Gbps) across %d workers\n",
			colorYellow("Interrupted", tty),
			colorCyan(formatSize(totalBytes), tty),
			colorBold(formatTime(elapsed), tty),
			formatFloat1(avgMBs), avgGbps, d.numWorkers)
		return
	}

	// Normal completion — replace live display with summary box
	d.moveUpAndClear()
	d.linesPrinted = 0

	termWidth := getTerminalWidth()
	innerW := termWidth - 6
	if innerW < 46 {
		innerW = 46
	}
	if innerW > 72 {
		innerW = 72
	}

	var lines []string
	lines = append(lines, d.boxTop(innerW, tty))
	lines = append(lines, d.boxLine(
		fmt.Sprintf("%s  %s — wrote %s in %s",
			colorBold(colorCyan("pushfill "+d.version, tty), tty),
			colorGreen("Done", tty),
			colorCyan(formatSize(totalBytes), tty),
			colorBold(formatTime(elapsed), tty)),
		innerW, tty))
	lines = append(lines, d.boxLine(
		fmt.Sprintf("%s %s %s across %d workers",
			colorMagenta("Average:", tty),
			colorCyan(formatFloat1(avgMBs)+" MB/s", tty),
			colorDim(fmt.Sprintf("(%.2f Gbps)", avgGbps), tty),
			d.numWorkers),
		innerW, tty))
	lines = append(lines, d.boxBottom(innerW, tty))

	output := strings.Join(lines, "\n") + "\n"
	fmt.Fprint(os.Stdout, output)
}

// ---------------------------------------------------------------------------------------
//
//	Box Drawing Helpers
//
// ---------------------------------------------------------------------------------------

// ---------------------------------------------------------------------------------------
func (d *Display) moveUpAndClear() {
	if d.linesPrinted > 0 {
		fmt.Fprintf(os.Stdout, "\033[%dA\033[J", d.linesPrinted)
	}
}

// ---------------------------------------------------------------------------------------
func (d *Display) getBar(fraction float64, width int, tty bool) string {
	if fraction < 0 {
		fraction = 0
	}
	if fraction > 1 {
		fraction = 1
	}
	filled := int(fraction * float64(width))
	empty := width - filled
	barFill := colorGreen(strings.Repeat("█", filled), tty)
	barEmpty := colorDim(strings.Repeat("░", empty), tty)
	return barFill + barEmpty
}

// ---------------------------------------------------------------------------------------
func (d *Display) boxLine(content string, innerW int, tty bool) string {
	visible := visibleLen(content)
	pad := innerW - visible
	if pad < 0 {
		pad = 0
	}
	border := colorDim("│", tty)
	return fmt.Sprintf("%s  %s%s  %s", border, content, strings.Repeat(" ", pad), border)
}

// ---------------------------------------------------------------------------------------
func (d *Display) boxTop(innerW int, tty bool) string {
	return colorDim("╭"+strings.Repeat("─", innerW+4)+"╮", tty)
}

// ---------------------------------------------------------------------------------------
func (d *Display) boxSep(innerW int, tty bool) string {
	return colorDim("├"+strings.Repeat("─", innerW+4)+"┤", tty)
}

// ---------------------------------------------------------------------------------------
func (d *Display) boxBottom(innerW int, tty bool) string {
	return colorDim("╰"+strings.Repeat("─", innerW+4)+"╯", tty)
}

// ---------------------------------------------------------------------------------------
//
//	Formatting Helpers
//
// ---------------------------------------------------------------------------------------

// ---------------------------------------------------------------------------------------
// formatSize formats a byte count as a human-readable string.
func formatSize(nbytes int64) string {
	if nbytes < 1024 {
		return fmt.Sprintf("%d B", nbytes)
	} else if nbytes < 1024*1024 {
		return fmt.Sprintf("%.1f KB", float64(nbytes)/1024)
	} else if nbytes < 1024*1024*1024 {
		return fmt.Sprintf("%.1f MB", float64(nbytes)/(1024*1024))
	} else if nbytes < 1024*1024*1024*1024 {
		return fmt.Sprintf("%.2f GB", float64(nbytes)/(1024*1024*1024))
	}
	return fmt.Sprintf("%.2f TB", float64(nbytes)/(1024*1024*1024*1024))
}

// ---------------------------------------------------------------------------------------
// formatFloat1 formats a float with 1 decimal place and comma-separated thousands.
// Replicates Python's "{:,.1f}" format.
func formatFloat1(f float64) string {
	s := fmt.Sprintf("%.1f", f)
	// Split on decimal point
	parts := strings.SplitN(s, ".", 2)
	integer := parts[0]
	decimal := ""
	if len(parts) > 1 {
		decimal = "." + parts[1]
	}

	// Handle negative
	negative := false
	if len(integer) > 0 && integer[0] == '-' {
		negative = true
		integer = integer[1:]
	}

	// Insert commas from right
	var result []byte
	for i, c := range integer {
		if i > 0 && (len(integer)-i)%3 == 0 {
			result = append(result, ',')
		}
		result = append(result, byte(c))
	}

	if negative {
		return "-" + string(result) + decimal
	}
	return string(result) + decimal
}

// ---------------------------------------------------------------------------------------
// formatTime formats seconds as h:mm:ss or m:ss.
func formatTime(seconds float64) string {
	if seconds < 0 {
		seconds = 0
	}
	s := int(seconds)
	if s < 3600 {
		return fmt.Sprintf("%d:%02d", s/60, s%60)
	}
	return fmt.Sprintf("%d:%02d:%02d", s/3600, (s%3600)/60, s%60)
}
